### PR TITLE
SONAR-6939 Ability to override the decimal scale of float measures

### PR DESCRIPTION
**Pull request by simonbrandhof**

### CODE REVIEW

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/DefaultFormulaData.java [hunk 1]
```diff
@@ -19,52 +19,51 @@
...
-import java.util.Collection;
-import java.util.List;
-
 /**
  * @since 1.11
+ * @deprecated since 5.2. Aggregation of measures is provided by {@link org.sonar.api.ce.measure.MeasureComputer}. {@link org.sonar.api.batch.Decorator}
+ * and {@link Formula} are no more supported.
  */
+@Deprecated
 public class DefaultFormulaData implements FormulaData {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/DefaultFormulaData.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/AverageFormula.java [hunk 2]
```diff
@@ -82,56 +63,10 @@ public AverageFormula setFallbackForMainMetric(Metric fallbackMetric) {
...
-    }
-
-    return result;
+    throw fail();
   }
 
-  private static boolean shouldDecorateResource(FormulaData data, FormulaContext context) {
-    return !MeasureUtils.hasValue(data.getMeasure(context.getTargetMetric()));
+  private static RuntimeException fail() {
+    throw new UnsupportedOperationException("Unsupported since version 5.2. Decorators and formulas are not used anymore for aggregation measures. Please use org.sonar.api.ce.measure.MeasureComputer.");
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/AverageFormula.java]: MINOR Split this 202 characters long line (which is greater than 180 authorized). [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00103)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/Measure.java [hunk 3]
```diff
@@ -322,27 +322,22 @@ public Integer getIntValue() {
...
    *
-   * @param v         the measure value
-   * @param precision the measure value precision
-   * @return the measure object instance
+   * @return {@code this}
+   * @deprecated in 5.3. The decimal scale is given by the metric, not by the measure. Anyway this parameter was enforced to 1 before version 5.3.
    */
-  public Measure<G> setValue(@Nullable Double v, int precision) {
+  @Deprecated
+  public Measure<G> setValue(@Nullable Double v, int decimalScale) {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/Measure.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/Measure.java [hunk 4]
```diff
@@ -43,7 +41,9 @@
 
   /**
    * Default precision when saving a float type metric
+   * @deprecated in 5.3. Decimal scale is provided by metric, not by measure.
    */
+  @Deprecated
   public static final int DEFAULT_PRECISION = 1;
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/Measure.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/MetricJsonWriter.java [hunk 5]
```diff
@@ -40,7 +40,8 @@
   public static final String FIELD_QUALITATIVE = "qualitative";
   public static final String FIELD_HIDDEN = "hidden";
   public static final String FIELD_CUSTOM = "custom";
-  public static final Set<String> OPTIONAL_FIELDS = ImmutableSet.of(FIELD_NAME, FIELD_DESCRIPTION, FIELD_DOMAIN, FIELD_DIRECTION, FIELD_QUALITATIVE, FIELD_HIDDEN, FIELD_CUSTOM);
+  public static final String FIELD_DECIMAL_SCALE = "decimalScale";
+  public static final Set<String> OPTIONAL_FIELDS = ImmutableSet.of(FIELD_NAME, FIELD_DESCRIPTION, FIELD_DOMAIN, FIELD_DIRECTION, FIELD_QUALITATIVE, FIELD_HIDDEN, FIELD_CUSTOM, FIELD_DECIMAL_SCALE);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/MetricJsonWriter.java]: MINOR Split this 198 characters long line (which is greater than 180 authorized). [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00103)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/Metric.java [hunk 6]
```diff
@@ -20,6 +20,7 @@
 package org.sonar.api.measures;
 
 import com.google.common.base.Function;
+import com.google.common.base.Objects;
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/Metric.java]: MINOR Remove this unused import 'com.google.common.base.Objects'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/DefaultFormulaContext.java [hunk 7]
```diff
@@ -19,37 +19,42 @@
...
 import org.sonar.api.measures.Metric;
 import org.sonar.api.resources.Resource;
 
 /**
  * @since 1.11
+ * @deprecated since 5.2. Aggregation of measures is provided by {@link org.sonar.api.ce.measure.MeasureComputer}. {@link org.sonar.api.batch.Decorator}
+ * and {@link Formula} are no more supported.
  */
+@Deprecated
 public class DefaultFormulaContext implements FormulaContext {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/DefaultFormulaContext.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/FormulaContext.java [hunk 8]
```diff
@@ -23,7 +23,10 @@
 
 /**
  * @since 1.11
+ * @deprecated since 5.2. Aggregation of measures is provided by {@link org.sonar.api.ce.measure.MeasureComputer}. {@link org.sonar.api.batch.Decorator}
+ * and {@link Formula} are no more supported.
  */
+@Deprecated
 public interface FormulaContext {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/FormulaContext.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/FormulaData.java [hunk 9]
```diff
@@ -23,7 +23,10 @@
 
 /**
  * @since 1.11
+ * @deprecated since 5.2. Aggregation of measures is provided by {@link org.sonar.api.ce.measure.MeasureComputer}. {@link org.sonar.api.batch.Decorator}
+ * and {@link Formula} are no more supported.
  */
+@Deprecated
 public interface FormulaData {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/FormulaData.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 11 issues:
* MINOR 5 minor
* INFO 6 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Method 'Measure.setValue(...)' is deprecated.](https://github.com/SonarSource/sonarqube/blob/58f7e24621c618f8210197ed7037b7b60c31524a/sonar-plugin-api/src/main/java/org/sonar/api/measures/Measure.java#L105) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
* MINOR [Method 'Measure.setValue(...)' is deprecated.](https://github.com/SonarSource/sonarqube/blob/58f7e24621c618f8210197ed7037b7b60c31524a/sonar-plugin-api/src/main/java/org/sonar/api/measures/Measure.java#L296) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues:
* MINOR 2 minor
* INFO 6 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Method 'Measure.setValue(...)' is deprecated.](https://github.com/SonarSource/sonarqube/blob/0364784cc9261694c06ade63d644b3fd83db90b0/sonar-plugin-api/src/main/java/org/sonar/api/measures/Measure.java#L105) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
* MINOR [Method 'Measure.setValue(...)' is deprecated.](https://github.com/SonarSource/sonarqube/blob/0364784cc9261694c06ade63d644b3fd83db90b0/sonar-plugin-api/src/main/java/org/sonar/api/measures/Measure.java#L296) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by julienlancelot** [comment]
LGTM

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues:
* MINOR 2 minor
* INFO 6 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Method 'Measure.setValue(...)' is deprecated.](https://github.com/SonarSource/sonarqube/blob/28fbbdea56ba6bf753819188b5bb34add3d525c1/sonar-plugin-api/src/main/java/org/sonar/api/measures/Measure.java#L105) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
* MINOR [Method 'Measure.setValue(...)' is deprecated.](https://github.com/SonarSource/sonarqube/blob/28fbbdea56ba6bf753819188b5bb34add3d525c1/sonar-plugin-api/src/main/java/org/sonar/api/measures/Measure.java#L296) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues:
* MINOR 2 minor
* INFO 6 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Method 'Measure.setValue(...)' is deprecated.](https://github.com/SonarSource/sonarqube/blob/28fbbdea56ba6bf753819188b5bb34add3d525c1/sonar-plugin-api/src/main/java/org/sonar/api/measures/Measure.java#L105) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
* MINOR [Method 'Measure.setValue(...)' is deprecated.](https://github.com/SonarSource/sonarqube/blob/28fbbdea56ba6bf753819188b5bb34add3d525c1/sonar-plugin-api/src/main/java/org/sonar/api/measures/Measure.java#L296) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Pull request closed without merging**